lib/core/versions/latest/DocumentComposer.ts

```diff
@@ -166,4 +171,123 @@ export default class DocumentComposer {
       }
     }
   }
+  /**
+   * Applies the given patches in order to the given DID Document.
+   * NOTE: Assumes no schema validation is needed.
+   */
+  private static applyPatchesToDidDocument (didDocument: DidDocumentModel, patches: any[]) {
+    // Loop through and apply all patches.
+    for (let patch of patches) {
+      DocumentComposer.applyPatchToDidDocument(didDocument, patch);
+    }
+  }
+
+  /**
+   * Applies the given patch to the given DID Document.
+   */
+  private static applyPatchToDidDocument (didDocument: DidDocumentModel, patch: any) {
+    if (patch.action === 'add-public-keys') {
+      const publicKeySet = new Set(didDocument.publicKey.map(key => key.id));
+
+      // Loop through all given public keys and add them if they don't exist already.
+      for (let publicKey of patch.publicKeys) {
+        if (!publicKeySet.has(publicKey.id)) {
+          // Add the controller property. This cannot be added by the client and can
+          // only be set by the server side
+          publicKey.controller = didDocument.id;
+          didDocument.publicKey.push(publicKey);
+        }
+      }
+    } else if (patch.action === 'remove-public-keys') {
+      const publicKeyMap = new Map(didDocument.publicKey.map(publicKey => [publicKey.id, publicKey]));
+
+      // Loop through all given public key IDs and delete them from the existing public key only if it is not a recovery key.
+      for (let publicKey of patch.publicKeys) {
+        const existingKey = publicKeyMap.get(publicKey);
+
+        // Deleting recovery key is NOT allowed.
+        if (existingKey !== undefined &&
+            existingKey.usage !== KeyUsage.recovery) {
+          publicKeyMap.delete(publicKey);
```

Comment:
Shouldn't this be a throw for the `else` case here? If we do want to throw, maybe the check belongs in the corresponding `validate ...` logic but I understand if we don't want to do the find in the document there.

Reply:
Concept of `usage` is no longer there and will be removed in my next round of refactoring (yes, there is at least one more round of refactoring...), it is no longer possible to remove a recovery key at all because the recovery key is not even in the document anymore. Will add a comment with issue number tracking removal of `usage`.